kafkaesk/retry.py

```diff
@@ -0,0 +1,216 @@
+    topic: str
+    partition: int
+    offset: int
+    timestamp: int
+    timestamp_type: int
+    key: Optional[str] = None
+    value: bytes
+    checksum: Optional[int] = None
+    serialized_key_size: int
+    serialized_value_size: int
+    headers: tuple
+
+    @classmethod
+    def from_consumer_record(cls, record: ConsumerRecord) -> "Record":
+        return cls(
+            topic=record.topic,
+            partition=record.partition,
+            offset=record.offset,
+            timestamp=record.timestamp,  # type: ignore
+            timestamp_type=record.timestamp_type,  # type: ignore
+            key=record.key,  # type: ignore
+            value=record.value,  # type: ignore
+            checksum=record.checksum,  # type: ignore
+            serialized_key_size=record.serialized_key_size,  # type: ignore
+            serialized_value_size=record.serialized_value_size,  # type: ignore
+            headers=record.headers,
+        )
+
+    def to_consumer_record(self) -> ConsumerRecord:
+        return ConsumerRecord(**self.dict())  # type: ignore
+
+
+class RetryInfo(BaseModel):
+    retry_count: int
+    retry_delay: int = 0
+    retry_timestamp: datetime.datetime
+    publish_timestamp: datetime.datetime
+    publish_topic: str
+    error: str
+
+
+class RetryMessage(BaseModel):
+    retry_info: RetryInfo
+    original_record: Record
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class FailureInfo(BaseModel):
+    retry_count: int
+    failure_timestamp: datetime.datetime
+    publish_topic: str
+    error: str
+
+
+class FailureMessage(BaseModel):
+    failure_info: FailureInfo
+    original_record: Record
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class RetryPolicy(ABC):
+    def __init__(self, no_retry_exceptions: Optional[List[Type[Exception]]] = None) -> None:
+        self._initialized = False
+        self._no_requeue_exceptions = no_retry_exceptions or []
+
+    async def initialize(self, app: "Application", subscription: "Subscription") -> None:
+
+        self._app = app
+        self._subscription = subscription
+
+        self._initialized = True
+
+    async def finalize(self) -> None:
+        self._initialized = False
+
+    def add_no_requeue_exception(self, exception: Type[Exception]) -> None:
+        if exception not in self._no_requeue_exceptions:
+            self._no_requeue_exceptions.append(exception)
+
+    def remove_no_requeue_exception(self, exception: Type[Exception]) -> None:
+        if exception in self._no_requeue_exceptions:
+            self._no_requeue_exceptions.remove(exception)
+
+    def get_no_requeue_exceptions(self) -> List[Type[Exception]]:
+        return self._no_requeue_exceptions
+
+    async def __call__(self, record: ConsumerRecord, error: Exception) -> None:
+        if self._initialized is not True:
+            raise RuntimeError("RetryPolicy is not initialized")
+
+        if self._should_requeue(record, error):
```

Comment:
maybe i am not seeing clearly but, this if can be modified to have only one error handler and not requeue and failure independently since its always one or the other, so the inferface to implement will be something like `async def handle_failure(...):`